### DISCUSSION TITLE
Lazy Loading for client side with server side generation

### DISCUSSION THREAD

**Discussion by GuyoST** [remix-run/remix#4341] [Q&A] [answered]
Hello every one, 

I was checking my project to reduce the javascript bundle size, and I I have component that can render other component depending on some conditions, like this for example : 

``` tsx
export function Embedder({val}} {
     switch(val) {
            case "a": return <ComponentA />
            case "b": return <ComponentB />
            case "c": return <ComponentC />
            default: return <Alert>ERROR</Alert>
      }
}
```

and of course I import component A B and C  : 

``` ts
import componentA from "./componentA"
import componentB from "./componentB"
import componentC from "./componentC"
```

but only 1 component will be displayed, but all 3 components are included in the javascript bundle.

I would like the component to be rendered server side to make the first load very fast, but, I would like on client side to lazy load the one which is actually used after the hydratation is done.

How should I do ?
What is the best practice ?

**Comment by sergiodxa** [comment, marked as answer]
You need to use a dynamic import combined with React.lazy, then your Embedder can render them and you can wrap Embedder in React.Suspense or wrap each item inside Embedder in React.Suspense.

Ensure you're using React 18 so you use suspense with SSR, if you're not and can't upgrade you'll have to wrap your Suspense components in a [ClientOnly component](https://github.com/sergiodxa/remix-utils#clientonly).

**Reply by GuyoST** [reply]
Hello thank you very much.

I could notice that I have React 18 in my project.
To use Lazy Loading and Suspense, I had to add `"module": "ESNext" in `tsconfig.json`

Maybe at the creation of a new project, this option should be set by default ?

**Reply by sergiodxa** [reply]
You shouldn't need to add ` "module": "ESNext"` to your tsconfig.json.

**Reply by lili21** [reply]
TS would yell about it
`Dynamic imports are only supported when the '--module' flag is set to 'es2020', 'es2022', 'esnext', 'commonjs', 'amd', 'system', 'umd', 'node16', or 'nodenext'.ts(1323)`

**Reply by jdnichollsc** [reply]
@sergiodxa ohh what's the way to use a dynamic import with RemixJS? :open_mouth:

**Reply by GuyoST** [reply]
> @sergiodxa ohh what's the way to use a dynamic import with RemixJS? 😮

Well you can check my first example :

``` ts
export function Embedder({val}} {
     switch(val) {
            case "a": return <ComponentA />
            case "b": return <ComponentB />
            case "c": return <ComponentC />
            default: return <Alert>ERROR</Alert>
      }
}
```
and of course I import component A B and C :

``` tsx
import componentA from "./componentA"
import componentB from "./componentB"
import componentC from "./componentC"
```

if I use normal import like this, all the component A, B and C will be sent to the browser, so bundle size will be larger.
With dynamic import instead of simple import, only the used component will be bundle and the size will be smaller.

Is it a good example to use Dynamic Imports ?
In this situation, it's not possible to use nested routes.

**Reply by jdnichollsc** [reply]
@GuyoST as you said, that's not a dynamic import, so I'm not sure why you share that code example again. I asked to Sergio because he commented about dynamic imports with React and RemixJS, I'm really interested to know more about that 😺

**Reply by sergiodxa** [reply]
@jdnichollsc I wrote an example and guide about this here https://sergiodxa.com/articles/lazy-load-react-components-in-remix
